File: src/forex_bot/core/storage.py
from __future__ import annotations

import json
import logging
import sqlite3
from collections import deque
from datetime import UTC, datetime
from pathlib import Path
from typing import TYPE_CHECKING, Any

import numpy as np

if TYPE_CHECKING:
    from ..features.talib_mixer import TALibStrategyGene
    from .config import Settings

logger = logging.getLogger(__name__)


class StrategyLedger:
    def __init__(self, db_path: str) -> None:
        path = Path(db_path)
        if not path.is_absolute():
            # Anchor relative paths to repo root (avoid CWD-dependent I/O errors).
            repo_root = Path(__file__).resolve().parents[3]
            path = repo_root / path
        self.db_path = str(path)
        self._init_db()

    def _init_db(self) -> None:
        db_path = Path(self.db_path)
        db_path.parent.mkdir(parents=True, exist_ok=True)
        try:
            conn = sqlite3.connect(self.db_path)
        except sqlite3.OperationalError as exc:
            msg = str(exc).lower()
            if "disk i/o" in msg or "disk io" in msg:
                # Attempt a clean rebuild of the ledger database.
                for suffix in ("", "-wal", "-shm"):
                    try:
                        (db_path.with_suffix(db_path.suffix + suffix)).unlink(missing_ok=True)
                    except Exception:
                        pass
                conn = sqlite3.connect(self.db_path)
            else:
                raise
        with conn:
            # Enable Write-Ahead Logging for concurrency
            conn.execute("PRAGMA journal_mode=WAL;")
            cursor = conn.cursor()

            cursor.execute("""
                CREATE TABLE IF NOT EXISTS alpha_strategies (
                    strategy_id TEXT PRIMARY KEY,
                    indicators TEXT,
                    params TEXT,
                    weights TEXT,
                    thresholds TEXT,
                    smc_flags TEXT,
                    metrics TEXT,
                    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                    last_used TIMESTAMP
                )
            """)

            cursor.execute("""
                CREATE TABLE IF NOT EXISTS trade_log (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    closed_at TIMESTAMP,
                    regime TEXT,
                    pnl REAL,
                    risk_pct REAL,
                    confidence REAL,
                    context TEXT
                )
            """)

            cursor.execute("""
                CREATE TABLE IF NOT EXISTS pending_intents (
                    ticket INTEGER PRIMARY KEY,
                    symbol TEXT,
                    direction TEXT,
                    volume REAL,
                    sl REAL,
                    tp REAL,
                    meta_risk_mult REAL,
                    timestamp TIMESTAMP DEFAULT CURRENT_TIMESTAMP
                )
            """)

            cursor.execute("""
                CREATE TABLE IF NOT EXISTS settings (
                    key TEXT PRIMARY KEY,
                    value TEXT
                )
            """)

            cursor.execute("""
                CREATE TABLE IF NOT EXISTS live_metrics (
                    symbol TEXT PRIMARY KEY,
                    timestamp TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                    returns_json TEXT,
                    sharpe REAL,
                    volatility REAL,
                    drawdown REAL,
                    position_count INTEGER,
                    equity REAL
                )
            """)

    def log_live_metrics(
        self,
        symbol: str,
        returns: list[float],
        sharpe: float,
        volatility: float,
        drawdown: float,
        position_count: int,
        equity: float,
    ) -> None:
        try:
            with sqlite3.connect(self.db_path) as conn:
                cursor = conn.cursor()
                returns_json = json.dumps(returns)
                cursor.execute(
                    """
                    INSERT OR REPLACE INTO live_metrics
                    (symbol, timestamp, returns_json, sharpe, volatility, drawdown, position_count, equity)
                    VALUES (?, CURRENT_TIMESTAMP, ?, ?, ?, ?, ?, ?)
                """,
                    (symbol, returns_json, sharpe, volatility, drawdown, position_count, equity),
                )
        except Exception as e:
            logger.error(f"Failed to log live metrics for {symbol}: {e}")

    def get_all_live_metrics(self) -> dict[str, dict[str, Any]]:
        try:
            with sqlite3.connect(self.db_path) as conn:
                conn.row_factory = sqlite3.Row
                cursor = conn.cursor()
                cursor.execute("SELECT * FROM live_metrics")
                rows = cursor.fetchall()

                all_metrics = {}
                for row in rows:
                    metrics = dict(row)
                    metrics["returns"] = json.loads(metrics["returns_json"])
                    del metrics["returns_json"]
                    all_metrics[metrics["symbol"]] = metrics
                return all_metrics
        except Exception:
            return {}

    def log_intent(
        self,
        ticket: int,
        symbol: str,
        direction: str,
        volume: float,
        sl: float,
        tp: float,
        meta_risk_mult: float,
    ) -> None:
        """Log an order intent immediately after placement for recovery."""
        try:
            with sqlite3.connect(self.db_path) as conn:
                cursor = conn.cursor()
                cursor.execute(
                    """
                    INSERT OR REPLACE INTO pending_intents (ticket, symbol, direction, volume, sl, tp, meta_risk_mult)
                    VALUES (?, ?, ?, ?, ?, ?, ?)
                """,
                    (ticket, symbol, direction, volume, sl, tp, meta_risk_mult),
                )
        except Exception as e:
            logger.error(f"Failed to log intent: {e}")

    def get_pending_intents(self) -> list[dict[str, Any]]:
        """Retrieve pending intents to reconcile with MT5 state."""
        try:
            with sqlite3.connect(self.db_path) as conn:
                conn.row_factory = sqlite3.Row
                cursor = conn.cursor()
                cursor.execute("SELECT * FROM pending_intents")
                rows = cursor.fetchall()
                return [dict(row) for row in rows]
        except Exception:
            return []

    def remove_intent(self, ticket: int) -> None:
        try:
            with sqlite3.connect(self.db_path) as conn:
                cursor = conn.cursor()
                cursor.execute("DELETE FROM pending_intents WHERE ticket = ?", (ticket,))
        except Exception as e:
            logger.error(f"Data persistence failed: {e}", exc_info=True)

    def save_alpha(self, gene: TALibStrategyGene) -> None:
        """Save a high-performing strategy gene to the ledger."""
        try:
            with sqlite3.connect(self.db_path) as conn:
                cursor = conn.cursor()

                indicators_json = json.dumps(gene.indicators)
                params_json = json.dumps(gene.params)
                weights_json = json.dumps(gene.weights)
                thresholds_json = json.dumps(
                    {"long": gene.long_threshold, "short": gene.short_threshold, "method": gene.combination_method}
                )
                smc_flags_json = json.dumps(
                    {
                        "use_ob": gene.use_ob,
                        "use_fvg": gene.use_fvg,
                        "use_liq_sweep": gene.use_liq_sweep,
                        "mtf_confirmation": gene.mtf_confirmation,
                        "use_premium_discount": getattr(gene, "use_premium_discount", False),
                        "use_inducement": getattr(gene, "use_inducement", False),
                    }
                )
                metrics_json = json.dumps(
                    {"fitness": gene.fitness, "sharpe": gene.sharpe_ratio, "win_rate": gene.win_rate}
                )

                cursor.execute(
                    """
                    INSERT OR REPLACE INTO alpha_strategies
                    (strategy_id, indicators, params, weights, thresholds, smc_flags, metrics, last_used)
                    VALUES (?, ?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)
                """,
                    (
                        gene.strategy_id,
                        indicators_json,
                        params_json,
                        weights_json,
                        thresholds_json,
                        smc_flags_json,
                        metrics_json,
                    ),
                )

                logger.info(f"Saved Alpha Strategy: {gene.strategy_id}")
        except Exception as e:
            logger.error(f"Failed to save alpha strategy: {e}")

    def load_top_alphas(self, limit: int = 10) -> list[TALibStrategyGene]:
        """Load top strategies sorted by fitness."""
        from ..features.talib_mixer import TALibStrategyGene  # Runtime import

        genes = []
        try:
            with sqlite3.connect(self.db_path) as conn:
                conn.row_factory = sqlite3.Row
                cursor = conn.cursor()

                cursor.execute("SELECT * FROM alpha_strategies")
                rows = cursor.fetchall()

                for row in rows:
                    try:
                        metrics = json.loads(row["metrics"])
                        fitness = metrics.get("fitness", 0.0)

                        thresholds = json.loads(row["thresholds"])
                        smc = json.loads(row["smc_flags"])

                        gene = TALibStrategyGene(
                            indicators=json.loads(row["indicators"]),
                            params=json.loads(row["params"]),
                            combination_method=thresholds.get("method", "weighted_vote"),
                            long_threshold=thresholds.get("long", 0.6),
                            short_threshold=thresholds.get("short", -0.6),
                            weights=json.loads(row["weights"]),
                            strategy_id=row["strategy_id"],
                            fitness=fitness,
                            sharpe_ratio=metrics.get("sharpe", 0.0),
                            win_rate=metrics.get("win_rate", 0.0),
                            use_ob=smc.get("use_ob", False),
                            use_fvg=smc.get("use_fvg", False),
                            use_liq_sweep=smc.get("use_liq_sweep", False),
                            mtf_confirmation=smc.get("mtf_confirmation", False),
                            use_premium_discount=smc.get("use_premium_discount", False),
                            use_inducement=smc.get("use_inducement", False),
                        )
                        genes.append(gene)
                    except Exception:
                        continue

                genes.sort(key=lambda x: x.fitness, reverse=True)
                return genes[:limit]

        except Exception as e:
            logger.error(f"Failed to load alphas: {e}")
            return []

    def log_trade(
        self,
        closed_at: datetime,
        regime: str,
        feature_vector: list[object] | None,
        model_votes: dict[str, object],
        policy: str,
        pnl: float,
        pnl_pct: float,
        risk_pct: float,
        slippage: float,
        news_context: dict[str, object],
        confidence: float,
        meta_risk_multiplier: float | None = None,
    ) -> None:
        try:
            with sqlite3.connect(self.db_path) as conn:
                cursor = conn.cursor()

                def default_serializer(obj: object) -> object:
                    if isinstance(obj, (np.integer, np.floating)):
                        return float(obj)
                    if isinstance(obj, np.ndarray):
                        return obj.tolist()
                    if hasattr(obj, "to_dict"):  # Pandas Series/DF
                        return obj.to_dict()
                    if hasattr(obj, "tolist"):  # Torch tensor
                        return obj.tolist()
                    return str(obj)

                context_json = json.dumps(
                    {
                        "features": feature_vector[:10] if feature_vector else [],  # Store only top 10 for space
                        "votes": model_votes,
                        "news": news_context,
                        "slippage": slippage,
                        "meta_risk_multiplier": meta_risk_multiplier,
                    },
                    default=default_serializer,
                )

                cursor.execute(
                    """
                    INSERT INTO trade_log (closed_at, regime, pnl, risk_pct, confidence, context)
                    VALUES (?, ?, ?, ?, ?, ?)
                """,
                    (closed_at, regime, pnl, risk_pct, confidence, context_json),
                )
                conn.commit()
        except Exception as e:
            logger.error(f"Failed to log trade: {e}", exc_info=True)
            raise

    def save_setting(self, key: str, value: object) -> None:
        try:
            with sqlite3.connect(self.db_path) as conn:
                cursor = conn.cursor()
                val_str = json.dumps(value)
                cursor.execute("INSERT OR REPLACE INTO settings (key, value) VALUES (?, ?)", (key, val_str))
        except Exception as e:
            logger.error(f"Data persistence failed: {e}", exc_info=True)

    def load_setting(self, key: str, default: object = None) -> object:
        try:
            with sqlite3.connect(self.db_path) as conn:
                cursor = conn.cursor()
                cursor.execute("SELECT value FROM settings WHERE key = ?", (key,))
                row = cursor.fetchone()
                if row:
                    return json.loads(row[0])
                return default
        except Exception:
            return default


class MetricsRecorder:
    def __init__(self, settings: Settings) -> None:
        self.db_path = settings.system.metrics_db_path
        self._init_db()

    def _init_db(self) -> None:
        """Initialize metrics database with proper schema."""
        try:
            import sqlite3

            conn = sqlite3.connect(self.db_path)
            cursor = conn.cursor()
            cursor.execute("""
                CREATE TABLE IF NOT EXISTS cycle_metrics (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    timestamp TEXT NOT NULL,
                    symbol TEXT,
                    equity REAL,
                    daily_pnl REAL,
                    session_trades INTEGER,
                    signal INTEGER,
                    confidence REAL,
                    position_size REAL,
                    data_json TEXT
                )
            """)
            cursor.execute("""
                CREATE INDEX IF NOT EXISTS idx_cycle_timestamp ON cycle_metrics(timestamp)
            """)
            cursor.execute("""
                CREATE INDEX IF NOT EXISTS idx_cycle_symbol ON cycle_metrics(symbol)
            """)
            conn.commit()
            conn.close()
            logger.debug(f"Metrics database initialized at {self.db_path}")
        except Exception as e:
            logger.error(f"Failed to initialize metrics database: {e}", exc_info=True)

    def log_cycle(self, data: dict[str, Any]) -> None:
        """Log trading cycle metrics to database."""
        try:
            import json
            import sqlite3
            from datetime import datetime

            conn = sqlite3.connect(self.db_path)
            cursor = conn.cursor()

            timestamp = data.get("timestamp", datetime.now(UTC).isoformat())
            symbol = data.get("symbol", "UNKNOWN")
            equity = data.get("equity", 0.0)
            daily_pnl = data.get("daily_pnl", 0.0)
            session_trades = data.get("session_trades", 0)
            signal = data.get("signal", 0)
            confidence = data.get("confidence", 0.0)
            position_size = data.get("position_size", 0.0)
            data_json = json.dumps(data, default=str)

            cursor.execute(
                """
                INSERT INTO cycle_metrics (
                    timestamp,
                    symbol,
                    equity,
                    daily_pnl,
                    session_trades,
                    signal,
                    confidence,
                    position_size,
                    data_json
                )
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
                (
                    timestamp,
                    symbol,
                    equity,
                    daily_pnl,
                    session_trades,
                    signal,
                    confidence,
                    position_size,
                    data_json,
                ),
            )

            conn.commit()
            conn.close()
        except Exception as e:
            logger.error(f"Failed to log cycle metrics: {e}", exc_info=True)


class RiskLedger:
    def __init__(self, max_events: int = 1000) -> None:
        self.events = deque(maxlen=max_events)

    def record(self, event_type: str, message: str, severity: str = "info", context: dict | None = None) -> None:
        event = {
            "timestamp": datetime.now(UTC).isoformat(),
            "type": event_type,
            "message": message,
            "severity": severity,
            "context": context or {},
        }
        self.events.append(event)

        if severity in ("warning", "error", "critical"):
            log_method = getattr(logger, severity, logger.info)
            log_method(f"Risk Event [{event_type}]: {message}")

File: crates/forex-core/src/storage.rs
use crate::config::Settings;
use anyhow::{Context, Result};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use std::collections::VecDeque;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use tracing::{debug, error, info, warn};

// use rusqlite::{params, Connection}; // will use fully qualified or add imports as needed

/// Represents a stored strategy gene (DTO for TALibStrategyGene)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SavedStrategy {
    pub strategy_id: String,
    pub indicators: JsonValue,
    pub params: JsonValue,
    pub weights: JsonValue,
    pub thresholds: JsonValue,
    pub smc_flags: JsonValue,
    pub metrics: JsonValue,
    pub last_used: Option<DateTime<Utc>>,
}

/// Main storage handler for strategies, trades, and intents
pub struct StrategyLedger {
    db_path: PathBuf,
}

impl StrategyLedger {
    pub fn new(db_path: impl AsRef<Path>, symbol: Option<&str>) -> Result<Self> {
        let mut path = db_path.as_ref().to_path_buf();

        // HPC FIX: Symbol-Isolated Databases
        if let Some(sym) = symbol {
            let stem = path
                .file_stem()
                .map(|s| s.to_string_lossy())
                .unwrap_or_default();
            path.set_file_name(format!("{}_{}.sqlite", stem, sym));
        }

        // Ensure absolute path
        if !path.is_absolute() {
            // In Rust, we usually assume CWD or relative to binary.
            // But if we want to mimic Python's repo_root anchoring, we might need logic.
            // For now, let's assume it references CWD or absolute path provided via config.
            // If relative, it remains relative to CWD.
        }

        let ledger = Self { db_path: path };
        ledger.init_db()?;
        Ok(ledger)
    }

    fn init_db(&self) -> Result<()> {
        if let Some(parent) = self.db_path.parent() {
            std::fs::create_dir_all(parent)?;
        }

        let conn = match rusqlite::Connection::open(&self.db_path) {
            Ok(c) => c,
            Err(e) => {
                // Handle disk I/O errors by attempting to clean up WAL files
                let msg = e.to_string().to_lowercase();
                if msg.contains("disk i/o") || msg.contains("disk io") {
                    warn!("Disk I/O error detected. cleaning up WAL/SHM files and retrying...");
                    // Try to remove -wal and -shm files
                    let _ = std::fs::remove_file(self.db_path.with_extension("sqlite-wal"));
                    let _ = std::fs::remove_file(self.db_path.with_extension("sqlite-shm"));
                    // Retry open
                    rusqlite::Connection::open(&self.db_path)?
                } else {
                    return Err(e.into());
                }
            }
        };

        // Enable Write-Ahead Logging and tuning
        conn.execute_batch(
            "PRAGMA journal_mode=WAL;
             PRAGMA wal_autocheckpoint=1000;
             PRAGMA synchronous=NORMAL;
             PRAGMA busy_timeout=5000;",
        )?;

        // Create tables
        conn.execute(
            "CREATE TABLE IF NOT EXISTS alpha_strategies (
                strategy_id TEXT PRIMARY KEY,
                indicators TEXT,
                params TEXT,
                weights TEXT,
                thresholds TEXT,
                smc_flags TEXT,
                metrics TEXT,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                last_used TIMESTAMP
            )",
            [],
        )?;

        conn.execute(
            "CREATE TABLE IF NOT EXISTS trade_log (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                closed_at TIMESTAMP,
                regime TEXT,
                pnl REAL,
                risk_pct REAL,
                confidence REAL,
                context TEXT
            )",
            [],
        )?;

        conn.execute(
            "CREATE TABLE IF NOT EXISTS pending_intents (
                ticket INTEGER PRIMARY KEY,
                symbol TEXT,
                direction TEXT,
                volume REAL,
                sl REAL,
                tp REAL,
                meta_risk_mult REAL,
                timestamp TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )",
            [],
        )?;

        conn.execute(
            "CREATE TABLE IF NOT EXISTS settings (
                key TEXT PRIMARY KEY,
                value TEXT
            )",
            [],
        )?;

        conn.execute(
            "CREATE TABLE IF NOT EXISTS live_metrics (
                symbol TEXT PRIMARY KEY,
                timestamp TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                returns_json TEXT,
                sharpe REAL,
                volatility REAL,
                drawdown REAL,
                position_count INTEGER,
                equity REAL
            )",
            [],
        )?;

        Ok(())
    }

    pub fn log_live_metrics(
        &self,
        symbol: &str,
        returns: &[f64],
        sharpe: f64,
        volatility: f64,
        drawdown: f64,
        position_count: i64,
        equity: f64,
    ) -> Result<()> {
        let conn = rusqlite::Connection::open(&self.db_path)?;
        let returns_json = serde_json::to_string(returns)?;

        conn.execute(
            "INSERT OR REPLACE INTO live_metrics
            (symbol, timestamp, returns_json, sharpe, volatility, drawdown, position_count, equity)
            VALUES (?1, CURRENT_TIMESTAMP, ?2, ?3, ?4, ?5, ?6, ?7)",
            rusqlite::params![
                symbol,
                returns_json,
                sharpe,
                volatility,
                drawdown,
                position_count,
                equity
            ],
        )
        .context("Failed to log live metrics")?;

        Ok(())
    }

    pub fn get_all_live_metrics(&self) -> Result<impl Iterator<Item = JsonValue>> {
        // Implementation note: In Rust we might return a dedicated Struct or HashMap.
        // Python returns dict[str, dict]. Let's keep it simple for now or return a Vec.
        // For efficiency, we just return a stub or Result<Vec<...>>.
        // Given complexity of returning iterator with open connection ref, let's fetch all into Vec.
        let conn = rusqlite::Connection::open(&self.db_path)?;
        let mut stmt = conn.prepare("SELECT * FROM live_metrics")?;

        // Mapping rows to JsonValue for flexible return, similar to Python dict
        let rows = stmt.query_map([], |row| {
            let symbol: String = row.get("symbol")?;
            let returns_json: String = row.get("returns_json")?;
            let returns: Vec<f64> = serde_json::from_str(&returns_json).unwrap_or_default();

            let metrics = serde_json::json!({
                "symbol": symbol,
                "returns": returns,
                "sharpe": row.get::<_, f64>("sharpe")?,
                "volatility": row.get::<_, f64>("volatility")?,
                "drawdown": row.get::<_, f64>("drawdown")?,
                "position_count": row.get::<_, i64>("position_count")?,
                "equity": row.get::<_, f64>("equity")?,
                "timestamp": row.get::<_, String>("timestamp")?,
            });
            Ok(metrics)
        })?;

        let mut results = Vec::new();
        for r in rows {
            results.push(r?);
        }
        Ok(results.into_iter())
    }

    pub fn save_alpha(&self, gene: &SavedStrategy) -> Result<()> {
        let conn = rusqlite::Connection::open(&self.db_path)?;

        // In Rust, we assume `gene` is already configured.
        // Logic to construct JSONs happens before call or `SavedStrategy` holds them as JsonValue.

        conn.execute(
            "INSERT OR REPLACE INTO alpha_strategies
            (strategy_id, indicators, params, weights, thresholds, smc_flags, metrics, last_used)
            VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, CURRENT_TIMESTAMP)",
            rusqlite::params![
                gene.strategy_id,
                gene.indicators.to_string(),
                gene.params.to_string(),
                gene.weights.to_string(),
                gene.thresholds.to_string(),
                gene.smc_flags.to_string(),
                gene.metrics.to_string()
            ],
        )?;

        info!("Saved Alpha Strategy: {}", gene.strategy_id);
        Ok(())
    }

    // Additional methods like log_intent, remove_intent, log_trade, etc. would follow similar patterns.
    // Implementing a few key ones for completeness.

    pub fn save_setting(&self, key: &str, value: &impl Serialize) -> Result<()> {
        let conn = rusqlite::Connection::open(&self.db_path)?;
        let val_str = serde_json::to_string(value)?;
        conn.execute(
            "INSERT OR REPLACE INTO settings (key, value) VALUES (?1, ?2)",
            rusqlite::params![key, val_str],
        )?;
        Ok(())
    }

    pub fn load_setting<T: for<'a> Deserialize<'a>>(&self, key: &str) -> Result<Option<T>> {
        let conn = rusqlite::Connection::open(&self.db_path)?;
        let mut stmt = conn.prepare("SELECT value FROM settings WHERE key = ?1")?;
        let mut rows = stmt.query(rusqlite::params![key])?;

        if let Some(row) = rows.next()? {
            let s: String = row.get(0)?;
            let val = serde_json::from_str(&s)?;
            Ok(Some(val))
        } else {
            Ok(None)
        }
    }
}

/// Metrics Recorder for cycle metrics
pub struct MetricsRecorder {
    db_path: PathBuf,
}

impl MetricsRecorder {
    pub fn new(settings: &Settings) -> Result<Self> {
        let path = PathBuf::from(&settings.system.metrics_db_path); // Assuming this field exists and is string/path
        let recorder = Self { db_path: path };
        recorder.init_db()?;
        Ok(recorder)
    }

    fn init_db(&self) -> Result<()> {
        if let Some(parent) = self.db_path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let conn = rusqlite::Connection::open(&self.db_path)?;

        conn.execute(
            "CREATE TABLE IF NOT EXISTS cycle_metrics (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                timestamp TEXT NOT NULL,
                symbol TEXT,
                equity REAL,
                daily_pnl REAL,
                session_trades INTEGER,
                signal INTEGER,
                confidence REAL,
                position_size REAL,
                data_json TEXT
            )",
            [],
        )?;

        conn.execute(
            "CREATE INDEX IF NOT EXISTS idx_cycle_timestamp ON cycle_metrics(timestamp)",
            [],
        )?;
        conn.execute(
            "CREATE INDEX IF NOT EXISTS idx_cycle_symbol ON cycle_metrics(symbol)",
            [],
        )?;

        debug!("Metrics database initialized at {:?}", self.db_path);
        Ok(())
    }

    pub fn log_cycle(&self, data: &serde_json::Map<String, JsonValue>) -> Result<()> {
        let conn = rusqlite::Connection::open(&self.db_path)?;

        let timestamp = data
            .get("timestamp")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(); // Should use chrono::Utc::now() if missing
        let symbol = data
            .get("symbol")
            .and_then(|v| v.as_str())
            .unwrap_or("UNKNOWN");
        let equity = data.get("equity").and_then(|v| v.as_f64()).unwrap_or(0.0);
        let daily_pnl = data
            .get("daily_pnl")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        let session_trades = data
            .get("session_trades")
            .and_then(|v| v.as_i64())
            .unwrap_or(0);
        let signal = data.get("signal").and_then(|v| v.as_i64()).unwrap_or(0);
        let confidence = data
            .get("confidence")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        let position_size = data
            .get("position_size")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        let data_json = serde_json::to_string(data)?;

        // If timestamp is empty, generate it
        let timestamp = if timestamp.is_empty() {
            Utc::now().to_rfc3339()
        } else {
            timestamp
        };

        conn.execute(
            "INSERT INTO cycle_metrics (
                timestamp, symbol, equity, daily_pnl, session_trades, signal, confidence, position_size, data_json
            ) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9)",
            rusqlite::params![
                timestamp, symbol, equity, daily_pnl, session_trades, signal, confidence, position_size, data_json
            ],
        )?;

        Ok(())
    }
}

/// Risk Event Ledger
pub struct RiskLedger {
    events: Arc<Mutex<VecDeque<RiskEvent>>>,
    max_events: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RiskEvent {
    pub timestamp: String, // ISO8601
    pub type_: String,
    pub message: String,
    pub severity: String,
    pub context: JsonValue,
}

impl RiskLedger {
    pub fn new(max_events: usize) -> Self {
        Self {
            events: Arc::new(Mutex::new(VecDeque::with_capacity(max_events))),
            max_events,
        }
    }

    pub fn record(
        &self,
        event_type: &str,
        message: &str,
        severity: &str,
        context: Option<JsonValue>,
    ) {
        let event = RiskEvent {
            timestamp: Utc::now().to_rfc3339(),
            type_: event_type.to_string(),
            message: message.to_string(),
            severity: severity.to_string(),
            context: context.unwrap_or(serde_json::json!({})),
        };

        if let Ok(mut lock) = self.events.lock() {
            if lock.len() >= self.max_events {
                lock.pop_front();
            }
            lock.push_back(event);
        }

        // Log to system logger
        match severity {
            "warning" | "WARN" => warn!("Risk Event [{}]: {}", event_type, message),
            "error" | "ERROR" => error!("Risk Event [{}]: {}", event_type, message),
            _ => info!("Risk Event [{}]: {}", event_type, message),
        }
    }
}
